src/coreclr/gc/gc.cpp

```diff
@@ -13219,16 +13248,123 @@ void region_free_list::sort_by_committed_and_age()
     for (heap_segment* region = new_head; region != nullptr; region = heap_segment_next (region))
     {
         heap_segment_prev_free_region (region) = prev;
-        assert ((prev == nullptr) || (compare_by_committed_and_age (prev, region) <= 0));
+        region_assert ((prev == nullptr) || (compare_by_committed_and_age (prev, region) <= 0));
         prev = region;
     }
     tail_free_region = prev;
 }
+
+int* free_list_snapshot::s_nullptr;
+int free_list_snapshot::s_dummy;
+
+VOLATILE(int32_t) free_list_snapshot::s_lock;
+
+free_list_snapshot* free_list_snapshot::s_buffer;
+
+int free_list_snapshot::s_counter_full;
+int free_list_snapshot::s_counter;
+
+void free_list_snapshot::init()
+{
+    s_nullptr = nullptr;
+    s_dummy = 0;
+    s_lock = 0;
+    s_counter = 0;
+    s_buffer = new free_list_snapshot[NUM_SNAPSHOTS];
+}
+
+void free_list_snapshot::fail()
+{
+    s_dummy = *s_nullptr;
+}
+
+void free_list_snapshot::record(heap_segment_snapshot* dst, heap_segment* src)
+{
```

Comment:
we might as well record the GC index here -

`VolatileLoadWithoutbarrier (&settings.gc_index)`